types/validator_set.go

```diff
@@ -269,9 +269,21 @@ func (vals *ValidatorSet) HasAddress(address []byte) bool {
 // GetByAddress returns an index of the validator with address and validator
 // itself (copy) if found. Otherwise, -1 and nil are returned.
 func (vals *ValidatorSet) GetByAddress(address []byte) (index int32, val *Validator) {
+	i, val := vals.GetByAddressMut(address)
+	if i == -1 {
+		return -1, nil
+	}
+	return i, val.Copy()
+}
+
+// GetByAddressMut returns an index of the validator with address and the
+// direct validator object if found. Mutations on this return value affect the validator set.
+// This method should be used by callers who will not mutate Val.
+// Otherwise, -1 and nil are returned.
+func (vals *ValidatorSet) GetByAddressMut(address []byte) (index int32, val *Validator) {
 	for idx, val := range vals.Validators {
 		if bytes.Equal(val.Address, address) {
-			return int32(idx), val.Copy()
+			return int32(idx), val
```

Comment:
Are 100% sure that no callsite will try to modify the validator structure we are returning here?

Reply:
Hmm... anyway, inside `Validator`, field `Address` is a slice, and field `PubKey` is an interface, so method `Copy()` in the original code wasn't really _copying_ these.

So, I like the idea of removing the `Copy()` because it makes the code less messy, but we need to make sure the other fields (`VotingPower` and `ProposerPriority`) are not modified at any callsite of `GetByAddressMut`.

Reply:
I've check the calls to the `GetByAddressMut` and seems not code invoking it are mutating the val returned. 

Afaiu, we're not eliminating the `Copy()` but just providing an alternative method to return a val that is not a `Copy()`

But I agree, that the `Copy()` is probably not doing a deep copy properly. One enhancement could be outside this PR is to add unit tests to ensure this fact doesn't have additional consequences.